### PR TITLE
TPC PID Skim Tree Creation

### PR DESCRIPTION
**Pull request by AnnalenaKa**
This pull request will add a check in the TPC PID Response class whether a tpc signal was present and the default values are changed to -999 for the resolution and number of sigma and to 0 for the expected signal.

In addition, a column with the expected signal is added to the TPC pid table, which is produced by the pid-tpc-full task.

Furthermore, the task DPG/Tasks/TPC/tpcSkimsTableCreator.cxx will allow to create a table, which will be used to train the neural network that is used for the multidimensional correction of the TPC response. The task is based on V0 and gamma decays plus TPC and TOF track selections at low momentum.

### CODE REVIEW

**Comment by AnnalenaKa** [comment]
Hi, 
@jezwilkinson , @ChSonnabend and @njacazio.
I have created the draft pull request adding the following changes:

1) Some hasTPC() checks in the TPCPIDResponse.h class. Furthermore, the default values were changed to -999 for the sigma and nsigma and to 0 for the expected signal. 

2) I had to add the expected signal to the tpc table which is produced by the pid-tpc-full task. It was stored for the TOF but not for the TPC and we need it for our tree creation. Otherwise we will have to copy parts of the tpc full task in order to get the expected signal. If Nicolo, you don't agree with that change I can revert that part and change the way we get the expected signal for the tree creation and get the response from the CCDB or a file in addition to running the pid-tpc-full task.

3) I added the skim tree creation task under DPG/Tasks/TPC. Please let me know if you want me to add some modifications.

File: Common/Core/PID/TPCPIDResponse.h [hunk 1]
```diff
@@ -89,6 +93,7 @@ class Response
   float mChargeFactor = 2.299999952316284f;
   float mMultNormalization = 11000.;
   bool mUseDefaultResolutionParam = true;
+  float nClNorm = 152.f;
 
   ClassDefNV(Response, 2);
```
Comment (jezwilkinson) [inline Common/Core/PID/TPCPIDResponse.h:98]: The ClassDef number should be incremented here for the new data member

File: Common/Tools/handleParamTPCResponse.cxx [hunk 2]
```diff
@@ -50,6 +50,7 @@ bool initOptionsAndParse(bpo::options_description& options, int argc, char* argv
     "paramMIP", bpo::value<float>()->default_value(50.f), "MIP parameter value")(
     "paramChargeFactor", bpo::value<float>()->default_value(2.299999952316284f), "Charge factor value")(
     "paramMultNormalization", bpo::value<float>()->default_value(11000.), "Multiplicity Normalization")(
+    "paramnClNormalization", bpo::value<float>()->default_value(152.), "Multiplicity Normalization")(
```
Comment (jezwilkinson) [inline Common/Tools/handleParamTPCResponse.cxx:53]: description: "Maximum nClusters for normalisation (159 for run 2, 152 for run 3)"

File: Common/TableProducer/PID/pidTPC.cxx [hunk 3]
```diff
@@ -183,6 +184,7 @@ struct tpcPid {
     reserveTable(pidAl, tablePIDAl);
 
     std::vector<float> network_prediction;
+    float nNclNormalization = response.GetNClNormalization();
```
Comment (wiechula) [inline Common/TableProducer/PID/pidTPC.cxx:187]:
```suggestion
    const float nNclNormalization = response.GetNClNormalization();
```

File: Common/TableProducer/PID/pidTPCFull.cxx [hunk 4]
```diff
@@ -214,7 +215,7 @@ struct tpcPidFull {
       const unsigned long prediction_size = output_dimensions * tracks_size;
 
       network_prediction = std::vector<float>(prediction_size * 9); // For each mass hypotheses
-
+      float nNclNormalization = response.GetNClNormalization();
```
Comment (wiechula) [inline Common/TableProducer/PID/pidTPCFull.cxx:218]:
```suggestion
      const float nNclNormalization = response.GetNClNormalization();
```

File: Common/TableProducer/PID/pidTPCML.h [hunk 5]
```diff
@@ -67,6 +69,8 @@ class Network
   std::vector<std::string> mOutputNames;
   std::vector<std::vector<int64_t>> mOutputShapes;
 
+  float nClNorm = 152.f;
```
Comment (wiechula) [inline Common/TableProducer/PID/pidTPCML.h:72]:
```suggestion
  const float nClNorm = 152.f;
```

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+
+  /// Kaon selection
+  template <typename C, typename V0>
+  bool selectionKaon(C const& collision, V0 const& v0)
+  {
+    // initialise dynamic variables
+    float alpha = v0.alpha();
+    float qt = v0.qtarm();
+    float cosPA = v0.v0cosPA(collision.posX(), collision.posY(), collision.posZ());
+    float q_K = {cutAPK0pinner * sqrt(abs(1 - ((alpha * alpha) / (cutAPK0alinner * cutAPK0alinner))))};
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:153]:
```suggestion
    const float alpha = v0.alpha();
    const float qt = v0.qtarm();
    const float cosPA = v0.v0cosPA(collision.posX(), collision.posY(), collision.posZ());
    const float q_K = {cutAPK0pinner * sqrt(abs(1 - ((alpha * alpha) / (cutAPK0alinner * cutAPK0alinner))))};
```
please also check below which other variables can be declared `const`. I think it is for most of the variables...

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+    return result;
+  };
+
+  /// Random downsampling trigger function using Tsalis/Hagedorn spectra fit (sqrt(s) = 62.4 GeV to 13 TeV)
+  /// as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
+  TRandom3* fRndm = new TRandom3(0);
+  bool downsampleTsalisCharged(double pt, float factor1Pt, double sqrts, double mass)
+  {
+    if (factor1Pt < 0.)
+      return true;
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:454]:
```suggestion
    if (factor1Pt < 0.) {
      return true;
    }
```

**Review by wiechula — CHANGES_REQUESTED** [review]
Please carefully check which variables can be declared const.

File: Common/Core/PID/TPCPIDResponse.h (hunk 1 above)
Comment (AnnalenaKa) [reply Common/Core/PID/TPCPIDResponse.h]: I set the ClassDef to 3.

File: Common/Core/PID/TPCPIDResponse.h [hunk 8]
```diff
@@ -122,7 +133,7 @@ inline float Response::GetExpectedSigma(const CollisionType& collision, const Tr
     const std::vector<double> values{1.f / dEdx, track.tgl(), std::sqrt(ncl), relReso, track.signed1Pt(), collision.multTPC() / mMultNormalization};
 
     const float reso = sqrt(pow(mResolutionParams[0], 2) * values[0] + pow(mResolutionParams[1], 2) * (values[2] * mResolutionParams[5]) * pow(values[0] / sqrt(1 + pow(values[1], 2)), mResolutionParams[2]) + values[2] * pow(values[3], 2) + pow(mResolutionParams[4] * values[4], 2) + pow(values[5] * mResolutionParams[6], 2) + pow(values[5] * (values[0] / sqrt(1 + pow(values[1], 2))) * mResolutionParams[7], 2)) * dEdx * mMIP;
-    reso >= 0.f ? resolution = reso : resolution = 0.f;
+    reso >= 0.f ? resolution = reso : resolution = -999.f;
```
[the remark this answers is not in the document]
Comment (AnnalenaKa) [reply Common/Core/PID/TPCPIDResponse.h]: I set the default for the resolution to 0 and check in the n sigma function for negative values and whether the track has a tpc signal. Like this we prevent a strange behaviour when using -999 for the resolution.

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+// void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
+// {
+//   std::vector<ConfigParamSpec> options{// runtime customisation goes here
+//                                        {"useV0", VariantType::Int, 0, {"Use V0 information for QA"}}};
+//   std::swap(workflowOptions, options);
+// }
+
+namespace o2::aod
+{
+namespace tpcskims
```
Comment (njacazio) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:48]: Here I have a suggestion that you might consider, feel free to ignore it as well. If you move the definition of the table to a header file you could a)  clean the actual code b) use it later in some lite version of the task that runs of skimmed data. This is how we do in the DPG.

**Review by njacazio — COMMENTED** [review]
Hi @AnnalenaKa this looks super nice indeed! I have just some minor comments/suggestions.
Did you estimate the compression factor of the skimmed data WRT the original AODs?

Also I'm wondering if you are interested into carrying out some down-sampling at some point.

**Comment by AnnalenaKa** [comment]
Hi, 
@jezwilkinson @wiechula and @njacazio 
I have included all your comments I received here or in the meeting. Please let me know if you have additional wishes or comments. If everyone is satisfied with the changes I would mark the request as ready for review.

**Comment by njacazio** [comment]
Thanks a lot @AnnalenaKa very nice, all good from me

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+    return result;
+  };
+
+  /// Random downsampling trigger function using Tsalis/Hagedorn spectra fit (sqrt(s) = 62.4 GeV to 13 TeV)
+  /// as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
+  TRandom3* fRndm = new TRandom3(0);
+  int downsampleTsalisCharged(double pt, double factor1Pt, double sqrts, double mass)
+  {
+    double prob = tsalisCharged(pt, mass, sqrts) * pt;
+    double probNorm = tsalisCharged(1., mass, sqrts);
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:302]:
```suggestion
    const double prob = tsalisCharged(pt, mass, sqrts) * pt;
    const double probNorm = tsalisCharged(1., mass, sqrts);
```

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+    int triggerMask = 0;
+    if ((fRndm->Rndm() * ((prob / probNorm) * pt * pt)) < factor1Pt)
+      triggerMask = 1;
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:305]:
```suggestion
    if ((fRndm->Rndm() * ((prob / probNorm) * pt * pt)) < factor1Pt) {
      triggerMask = 1;
      }
```

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+  /// Random downsampling trigger function using Tsalis/Hagedorn spectra fit (sqrt(s) = 62.4 GeV to 13 TeV)
+  /// as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
+  TRandom3* fRndm = new TRandom3(0);
+  bool downsampleTsalisCharged(double pt, float factor1Pt, double sqrts, double mass)
+  {
+    if (factor1Pt < 0.) {
+      return true;
+    }
+    double prob = tsalisCharged(pt, mass, sqrts) * pt;
+    double probNorm = tsalisCharged(1., mass, sqrts);
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:437]:
```suggestion
    const double prob = tsalisCharged(pt, mass, sqrts) * pt;
    const double probNorm = tsalisCharged(1., mass, sqrts);
```

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
+    const double mass = o2::track::pid_constants::sMasses[id];
+    const double bg = p / mass;
+    const int multTPC = collision.multTPC();
+
+    const float alpha = v0.alpha();
+    const float qt = v0.qtarm();
+    const float cosPA = v0.v0cosPA(collision.posX(), collision.posY(), collision.posZ());
+    const float pT = v0.pt();
+
+    double pseudoRndm = track.pt() * 1000. - (long)(track.pt() * 1000);
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:260]:
```suggestion
    const double pseudoRndm = track.pt() * 1000. - (long)(track.pt() * 1000);
```

File: DPG/Tasks/TPC/tpcSkimsTableCreator.cxx [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+  void fillSkimmedTPCTOFTable(T const& track, C const& collision, const float nSigmaTPC, const float nSigmaTOF, const float dEdxExp, const o2::track::PID::ID id, double dwnSmplFactor)
+  {
+
+    const double ncl = track.tpcNClsFound();
+    const double p = track.tpcInnerParam();
+    const double mass = o2::track::pid_constants::sMasses[id];
+    const double bg = p / mass;
+    const int multTPC = collision.multTPC();
+
+    double pseudoRndm = track.pt() * 1000. - (long)(track.pt() * 1000);
```
Comment (wiechula) [inline DPG/Tasks/TPC/tpcSkimsTableCreator.cxx:456]:
```suggestion
    const double pseudoRndm = track.pt() * 1000. - (long)(track.pt() * 1000);
```

**Review by wiechula — CHANGES_REQUESTED** [review]
Some more extremely minor suggestions.
Also I would prefer to explicitly use the `std::` math functions:
```
pow,TMath::Power -> std::pow
abs,TMath::Abs -> std::abs
sqrt -> std::sqrt
```
etc.

**Comment by jezwilkinson** [comment]
All looks good to me, thanks for the updates!

**Review by mfaggin — DISMISSED** [review]
I see that this PR was hugely discussed already, green light from my side then.

**Review by wiechula — DISMISSED** [review]
Thanks a lot @AnnalenaKa looks good!

**Comment by AnnalenaKa** [comment]
> I would just wait #1025 to be merged

Hi @njacazio I saw this pull request was merged. Could we proceed with the review now to merge this PR as well?

**Comment by njacazio** [comment]
> > I would just wait #1025 to be merged
> 
> Hi @njacazio I saw this pull request was merged. Could we proceed with the review now to merge this PR as well?

Sure, for me it's already fine! @jgrosseo can we have this merged?

**Comment by njacazio** [comment]
I fixed a conflict brought in by a recent PR merge that prevented the merge

**Review by njacazio — APPROVED** [review]

**Review by ddobrigk — APPROVED** [review]
Stalled PR, looks ok and reviewed, approved

**Review by iarsene — APPROVED** [review]

**Pull request merged**